packages/devtools_app/test/shared/disconnect_observer_test.dart

```diff
@@ -0,0 +1,147 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:devtools_app/devtools_app.dart';
+import 'package:devtools_app/src/framework/disconnect_observer.dart';
+import 'package:devtools_app/src/shared/framework_controller.dart';
+import 'package:devtools_app_shared/shared.dart';
+import 'package:devtools_app_shared/ui.dart';
+import 'package:devtools_app_shared/utils.dart';
+import 'package:devtools_test/devtools_test.dart';
+import 'package:devtools_test/helpers.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+void main() {
+  group('DisconnectObserver', () {
+    late FakeServiceConnectionManager fakeServiceConnectionManager;
+
+    setUp(() {
+      fakeServiceConnectionManager = FakeServiceConnectionManager();
+      setGlobal(ServiceConnectionManager, fakeServiceConnectionManager);
+      setGlobal(FrameworkController, FrameworkController());
+      setGlobal(OfflineDataController, OfflineDataController());
+      setGlobal(IdeTheme, IdeTheme());
+    });
+
+    Future<void> pumpDisconnectObserver(WidgetTester tester) async {
+      await tester.pumpWidget(
+        wrap(
+          Builder(
+            builder: (context) {
+              return DisconnectObserver(
+                routerDelegate: DevToolsRouterDelegate.of(context),
+                child: const Placeholder(),
+              );
+            },
+          ),
+        ),
+      );
+      await tester.pumpAndSettle();
+    }
+
+    void verifyObserverState(
+      WidgetTester tester, {
+      required bool connected,
+      required bool showingOverlay,
+    }) {
+      final DisconnectObserverState state =
+          tester.state(find.byType(DisconnectObserver));
+      expect(state.currentConnectionState.connected, connected);
+      expect(
+        state.currentDisconnectedOverlay,
+        showingOverlay ? isNotNull : isNull,
+      );
+      expect(
+        find.text('Disconnected'),
+        showingOverlay ? findsOneWidget : findsNothing,
+      );
+      expect(
+        find.byType(ConnectToNewAppButton),
+        showingOverlay && !isEmbedded() ? findsOneWidget : findsNothing,
+      );
+      expect(
+        find.text('Run a new debug session to reconnect'),
+        showingOverlay && isEmbedded() ? findsOneWidget : findsNothing,
+      );
+      expect(
+        find.text('Review recent data (offline)'),
+        showingOverlay && offlineDataController.offlineDataJson.isNotEmpty
+            ? findsOneWidget
+            : findsNothing,
+      );
+    }
+
+    testWidgets(
+      'initiailized in a disconnected state',
```

Comment:
nit: typo (here and below) - can be fixed in follow up if we don't want to wait for ci bots again though